### PR TITLE
Allow upper case letters in proc names

### PR DESCRIPTION
**Pull request by alonre24**
#1369

### CODE REVIEW

File: src/procedures/procedure.c [hunk 1]
```diff
@@ -58,7 +59,10 @@ ProcedureCtx *ProcCtxNew(const char *name,
 
 ProcedureCtx *Proc_Get(const char *proc_name) {
 	if(!__procedures) return NULL;
-	ProcGenerator gen = raxFind(__procedures, (unsigned char *)proc_name, strlen(proc_name));
+	char *proc_name_lower = rm_malloc(strlen(proc_name)+1);
+    for (int i=0; i<strlen(proc_name); i++) proc_name_lower[i] = tolower(proc_name[i]);
```
Comment (swilly22) [inline src/procedures/procedure.c:63]: Add  `src/util/strutil.h` and `src/util/strutil.c`
Migrate `_tolower` and rename `str_tolower) function from `src/arithmetic/repository.c` to `strutil.c`
and use it.

use this new function in:
`src/arithmetic/func_desc.c`, `src/arithmetic/repository.c` `src/arithmetic/string_funcs/string_funcs.c`

File: src/procedures/procedure.c [hunk 2]
```diff
@@ -58,7 +59,10 @@ ProcedureCtx *ProcCtxNew(const char *name,
...
+	ProcGenerator gen = raxFind(__procedures, (unsigned char *)proc_name_lower, strlen(proc_name));
```
Comment (swilly22) [inline src/procedures/procedure.c:64]: No need to compute `strlen(proc_name)` twice, save first call to a local variable.

File: src/procedures/procedure.c (hunk 1 above)
Comment (swilly22) [inline src/procedures/procedure.c:63]: Ident.

File: src/procedures/procedure.c [hunk 3]
```diff
@@ -58,7 +59,10 @@ ProcedureCtx *ProcCtxNew(const char *name,
 
 ProcedureCtx *Proc_Get(const char *proc_name) {
 	if(!__procedures) return NULL;
-	ProcGenerator gen = raxFind(__procedures, (unsigned char *)proc_name, strlen(proc_name));
+	char *proc_name_lower = rm_malloc(strlen(proc_name)+1);
```
Comment (swilly22) [inline src/procedures/procedure.c:62]: `rm_malloc(strlen(proc_name)+1);`
`rm_malloc(strlen(proc_name) + 1);`

File: src/procedures/procedure.c (hunk 1 above)
Comment (swilly22) [inline src/procedures/procedure.c:63]: `for (int i=0; i<strlen(proc_name); i++)`
`for (int i = 0; i < strlen(proc_name); i++)`

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/util/strutil.h [hunk 4]
```diff
@@ -0,0 +1,14 @@
+//
+// Created by alon on 07/10/2020.
+//
```
Comment (swilly22) [inline src/util/strutil.h:3]: Please follow the standard copy rights comment we use in all of our files.

File: src/util/strutil.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+#ifndef REDISGRAPH_STRUTIL_H
+#define REDISGRAPH_STRUTIL_H
```
Comment (swilly22) [inline src/util/strutil.h:6]: Replace with `#pragma once`

File: src/util/strutil.h [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+void str_tolower(const char *str, char *lower, short *lower_len);
```
Comment (swilly22) [inline src/util/strutil.h:8]: Add a short comment explaining the function and its arguments.

File: src/util/strutil.h [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+void str_toupper(const char *str, char *lower, short *lower_len);
```
Comment (swilly22) [inline src/util/strutil.h:10]: Add a short comment explaining the function and its arguments.
Also please rename the arguments accordingly.

File: src/util/strutil.h [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+
+
```
Comment (swilly22) [inline src/util/strutil.h:13]: Remove extra lines

File: src/util/strutil.h [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+#endif //REDISGRAPH_STRUTIL_H
```
Comment (swilly22) [inline src/util/strutil.h:14]: Remove, not needed when using `#pragma once`

File: src/util/strutil.c [hunk 10]
```diff
@@ -0,0 +1,33 @@
+//
+// Created by alon on 07/10/2020.
+//
```
Comment (swilly22) [inline src/util/strutil.c:3]: See our standard copyright comment

File: src/util/strutil.c [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+#include "strutil.h"
+#include <string.h>
+#include <ctype.h>
+#include <assert.h>
+
+
```
Comment (swilly22) [inline src/util/strutil.c:9]: Please remove extra new line

File: src/util/strutil.c [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+void str_tolower(const char *str, char *lower, short *lower_len) {
+    size_t str_len = strlen(str);
+    /* Avoid overflow. */
+    assert(*lower_len > str_len);
+
+    /* Update lower len*/
```
Comment (swilly22) [inline src/util/strutil.c:15]: space between closing comment `*/` and the comment itself.
`/* Update lower len */`

File: src/util/strutil.c [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+//
+#include "strutil.h"
+#include <string.h>
+#include <ctype.h>
+#include <assert.h>
+
+
+void str_tolower(const char *str, char *lower, short *lower_len) {
+    size_t str_len = strlen(str);
+    /* Avoid overflow. */
```
Comment (swilly22) [inline src/util/strutil.c:12]: Usually a single line comment uses: `//` instead of `/**/`

File: src/util/strutil.c [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    assert(*upper_len > str_len);
+
+    /* Update lower len*/
+    *upper_len = str_len;
+    size_t i = 0;
+    for(; i < str_len; i++) upper[i] = toupper(str[i]);
+    upper[i] = 0;
+}
```
Comment (swilly22) [inline src/util/strutil.c:33]: files end with a single empty new line.

File: src/procedures/procedure.c [hunk 15]
```diff
@@ -22,18 +22,28 @@ static void _procRegister(const char *procedure, ProcGenerator gen) {
 // Register procedures.
 void Proc_Register() {
 	__procedures = raxNew();
-	_procRegister("db.labels", Proc_LabelsCtx);
-	_procRegister("db.propertykeys", Proc_PropKeysCtx);
-	_procRegister("db.relationshiptypes", Proc_RelationsCtx);
+	char proc_name_lower[128] = {0};
+	short proc_name_lower_len = 128;
+    str_tolower("db.labels", &proc_name_lower[0], &proc_name_lower_len);
+	_procRegister(proc_name_lower, Proc_LabelsCtx);
```
Comment (swilly22) [inline src/procedures/procedure.c]: Please note indentation throughout this file

File: src/procedures/procedure.c [hunk 16]
```diff
@@ -22,18 +22,28 @@ static void _procRegister(const char *procedure, ProcGenerator gen) {
...
+    str_tolower("db.propertyKeys", &proc_name_lower[0], &proc_name_lower_len);
+    _procRegister(proc_name_lower, Proc_PropKeysCtx);
+    str_tolower("db.relationshipTypes", &proc_name_lower[0], &proc_name_lower_len);
+	_procRegister(proc_name_lower, Proc_RelationsCtx);
```
Comment (swilly22) [inline src/procedures/procedure.c]: the conversion to lower case should be done within `_procRegister`. much easier and cleaner.

File: src/procedures/procedure.c [hunk 17]
```diff
@@ -59,9 +69,10 @@ ProcedureCtx *ProcCtxNew(const char *name,
 
 ProcedureCtx *Proc_Get(const char *proc_name) {
 	if(!__procedures) return NULL;
-	char *proc_name_lower = rm_malloc(strlen(proc_name)+1);
-    for (int i=0; i<strlen(proc_name); i++) proc_name_lower[i] = tolower(proc_name[i]);
-	ProcGenerator gen = raxFind(__procedures, (unsigned char *)proc_name_lower, strlen(proc_name));
+	short proc_name_len = strlen(proc_name);
+	char *proc_name_lower = rm_malloc(proc_name_len + 1);
+    str_tolower(proc_name, proc_name_lower, &proc_name_len);
```
Comment (swilly22) [inline src/procedures/procedure.c]: Indentation.

File: src/arithmetic/string_funcs/string_funcs.c [hunk 18]
```diff
@@ -146,17 +135,17 @@ SIValue AR_TOLOWER(SIValue *argv, int argc) {
 	char *original = argv[0].stringval;
 	short lower_len = strlen(original);
 	char *lower = rm_malloc((lower_len + 1) * sizeof(char));
-	_toLower(original, lower, lower_len);
+    str_tolower(original, lower, &lower_len);
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c]: Indentation

File: src/arithmetic/string_funcs/string_funcs.c [hunk 19]
```diff
@@ -146,17 +135,17 @@ SIValue AR_TOLOWER(SIValue *argv, int argc) {
...
 
 /* returns the original string in uppercase. */
 SIValue AR_TOUPPER(SIValue *argv, int argc) {
 	if(SIValue_IsNull(argv[0])) return SI_NullVal();
 	char *original = argv[0].stringval;
-	size_t upper_len = strlen(original);
+	short upper_len = strlen(original);
 	char *upper = rm_malloc((upper_len + 1) * sizeof(char));
-	_toUpper(original, upper, upper_len);
+    str_toupper(original, upper, &upper_len);
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c]: Indentation

File: src/arithmetic/string_funcs/string_funcs.c [hunk 20]
```diff
@@ -146,17 +135,17 @@ SIValue AR_TOLOWER(SIValue *argv, int argc) {
...
+    str_tolower(original, lower, &lower_len);
 	return SI_TransferStringVal(lower);
 }
 
 /* returns the original string in uppercase. */
 SIValue AR_TOUPPER(SIValue *argv, int argc) {
 	if(SIValue_IsNull(argv[0])) return SI_NullVal();
 	char *original = argv[0].stringval;
-	size_t upper_len = strlen(original);
+	short upper_len = strlen(original);
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:146]: Why the change from `size_t` to `short` ?

File: src/arithmetic/repository.c [hunk 21]
```diff
@@ -35,7 +23,7 @@ void Agg_RegisterFunc(const char *name, AggFuncInit f) {
 	__agg_initRegistry();
 	char lower_func_name[32] = {0};
 	short lower_func_name_len = 32;
-	_toLower(name, &lower_func_name[0], &lower_func_name_len);
+    str_tolower(name, &lower_func_name[0], &lower_func_name_len);
```
Comment (swilly22) [inline src/arithmetic/repository.c]: Indentation

File: src/arithmetic/repository.c [hunk 22]
```diff
@@ -55,7 +43,7 @@ void Agg_GetFunc(const char *name, bool distinct, AggCtx **ctx) {
 
 	char lower_func_name[32] = {0};
 	short lower_func_name_len = 32;
-	_toLower(name, &lower_func_name[0], &lower_func_name_len);
+    str_tolower(name, &lower_func_name[0], &lower_func_name_len);
```
Comment (swilly22) [inline src/arithmetic/repository.c]: Indentation

File: src/arithmetic/func_desc.c [hunk 23]
```diff
@@ -46,7 +34,7 @@ AR_FuncDesc *AR_FuncDescNew(const char *name, AR_Func func, uint min_argc, uint
 void AR_RegFunc(AR_FuncDesc *func) {
 	char lower_func_name[32] = {0};
 	short lower_func_name_len = 32;
-	_toLower(func->name, &lower_func_name[0], &lower_func_name_len);
+    str_tolower(func->name, &lower_func_name[0], &lower_func_name_len);
```
Comment (swilly22) [inline src/arithmetic/func_desc.c]: Indentation

File: src/arithmetic/func_desc.c [hunk 24]
```diff
@@ -55,7 +43,7 @@ void AR_RegFunc(AR_FuncDesc *func) {
 AR_FuncDesc *AR_GetFunc(const char *func_name) {
 	char lower_func_name[32] = {0};
 	short lower_func_name_len = 32;
-	_toLower(func_name, &lower_func_name[0], &lower_func_name_len);
+    str_tolower(func_name, &lower_func_name[0], &lower_func_name_len);
```
Comment (swilly22) [inline src/arithmetic/func_desc.c]: Indentation

**Review by swilly22 — CHANGES_REQUESTED** [review]
Please see comments.

File: src/arithmetic/string_funcs/string_funcs.c (hunk 20 above)
Comment (alonre24) [reply src/arithmetic/string_funcs/string_funcs.c:146]: because I saw that "short" is used in every other place for storing the string length. Is it matter?

File: src/procedures/procedure.c (hunk 16 above)
Comment (alonre24) [reply src/procedures/procedure.c]: I agree, but in this case the function should return the new string instead of void right? so should I add another function or change the code in every place that uses str_tolower?

File: src/arithmetic/string_funcs/string_funcs.c (hunk 20 above)
Comment (swilly22) [reply src/arithmetic/string_funcs/string_funcs.c:146]: Yeah sort and size_t are probably of different sizes, also `strlen` returns `size_t`

File: src/procedures/procedure.c (hunk 16 above)
Comment (swilly22) [reply src/procedures/procedure.c]: Something along these lines:

```
_procRegister(proc_name, proc) {
char proc_name_lower[128];
uint proc_name_lower_len ;
str_tolower(proc_name, &proc_name_lower, &proc_name_lower_len);
...
}
```

File: src/util/strutil.h [hunk 25]
```diff
@@ -0,0 +1,15 @@
+/*
+* Copyright 2018-2020 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#pragma once
+
+#include <glob.h>
```
Comment (swilly22) [inline src/util/strutil.h:9]: What is this include for?
if it is needed in the `c` file please move it there.

File: src/procedures/procedure.c [hunk 26]
```diff
@@ -10,12 +10,17 @@
...
+#include "../util/strutil.h"
 #include "../graph/graphcontext.h"
 
 static rax *__procedures = NULL;
 
 static void _procRegister(const char *procedure, ProcGenerator gen) {
-	raxInsert(__procedures, (unsigned char *)procedure, strlen(procedure), gen, NULL);
+	char lower_proc_name[128];
+	size_t lower_proc_name_len = 128;
+	str_tolower(procedure, &lower_proc_name[0], &lower_proc_name_len);
```
Comment (swilly22) [inline src/procedures/procedure.c:21]: I think `&lower_proc_name[0]` can be replaced with just `lower_proc_name`
unless we get a compiler warning / error complaining that `char lower_proc_name[128]` isn't `char*`

File: src/procedures/procedure.c [hunk 27]
```diff
@@ -10,12 +10,17 @@
...
 #include "../util/arr.h"
 #include "../util/rmalloc.h"
+#include "../util/strutil.h"
 #include "../graph/graphcontext.h"
 
 static rax *__procedures = NULL;
 
 static void _procRegister(const char *procedure, ProcGenerator gen) {
-	raxInsert(__procedures, (unsigned char *)procedure, strlen(procedure), gen, NULL);
+	char lower_proc_name[128];
```
Comment (swilly22) [inline src/procedures/procedure.c:19]: Rename `lower_proc_name` to `lowercase_proc_name`

File: src/procedures/procedure.c [hunk 28]
```diff
@@ -10,12 +10,17 @@
...
+	size_t lower_proc_name_len = 128;
+	str_tolower(procedure, &lower_proc_name[0], &lower_proc_name_len);
+	raxInsert(__procedures, (unsigned char *)lower_proc_name, strlen(procedure),
```
Comment (swilly22) [inline src/procedures/procedure.c:22]: it's a shame to recompute the procedure name length `lower_proc_name_len` already holds that value...

File: src/procedures/procedure.c [hunk 29]
```diff
@@ -10,12 +10,17 @@
...
 #include "../util/rmalloc.h"
+#include "../util/strutil.h"
 #include "../graph/graphcontext.h"
 
 static rax *__procedures = NULL;
 
 static void _procRegister(const char *procedure, ProcGenerator gen) {
-	raxInsert(__procedures, (unsigned char *)procedure, strlen(procedure), gen, NULL);
+	char lower_proc_name[128];
+	size_t lower_proc_name_len = 128;
```
Comment (swilly22) [inline src/procedures/procedure.c:20]: Rename `lower_proc_name_len` to `lowercase_proc_name_len`

File: src/procedures/procedure.c [hunk 30]
```diff
@@ -58,7 +63,11 @@ ProcedureCtx *ProcCtxNew(const char *name,
 
 ProcedureCtx *Proc_Get(const char *proc_name) {
 	if(!__procedures) return NULL;
-	ProcGenerator gen = raxFind(__procedures, (unsigned char *)proc_name, strlen(proc_name));
+	size_t proc_name_len = strlen(proc_name);
+	char *proc_name_lower = rm_malloc(proc_name_len + 1);
```
Comment (swilly22) [inline src/procedures/procedure.c:67]: Come to think of it, we can do:
`char proc_name_lower[proc_name_len+1]`

File: src/procedures/procedure.c [hunk 31]
```diff
@@ -58,7 +63,11 @@ ProcedureCtx *ProcCtxNew(const char *name,
...
+	str_tolower(proc_name, &proc_name_lower[0], &proc_name_len);
```
Comment (swilly22) [inline src/procedures/procedure.c:68]: `&proc_name_lower[0]` is simply `proc_name_lower`

File: src/procedures/procedure.c [hunk 32]
```diff
@@ -58,7 +63,11 @@ ProcedureCtx *ProcCtxNew(const char *name,
...
+	ProcGenerator gen = raxFind(__procedures, (unsigned char *) proc_name_lower, proc_name_len);
+	rm_free(proc_name_lower);
```
Comment (swilly22) [inline src/procedures/procedure.c:70]: if we move to stack allocation, this needs to be removed.

File: src/procedures/procedure.c [hunk 33]
```diff
@@ -124,7 +132,8 @@ bool Procedure_ContainsOutput(const ProcedureCtx *proc, const char *output) {
 
 bool Proc_ReadOnly(const char *proc_name) {
 	assert(__procedures);
-	ProcGenerator gen = raxFind(__procedures, (unsigned char *)proc_name, strlen(proc_name));
+	ProcGenerator gen = raxFind(__procedures, (unsigned char *)proc_name,
+							 strlen(proc_name));
```
Comment (swilly22) [inline src/procedures/procedure.c:136]: ` strlen(proc_name));` should be ~2 tabs aligned, landing below `__procedures`

**Review by swilly22 — CHANGES_REQUESTED** [review]
getting there, just a few more comments.

File: deps/GraphBLAS/README.md [hunk 34]
```diff
@@ -3,7 +3,7 @@
 SuiteSparse:GraphBLAS, Timothy A. Davis, (c) 2017-2020, All Rights Reserved.
 http://suitesparse.com   See GraphBLAS/Doc/License.txt for license.
 
-VERSION 3.2.0, Feb 20, 2020 GB_control used for size reduction.
+VERSION 3.2.0, Feb 20, 2020
```
Comment (swilly22) [inline deps/GraphBLAS/README.md:6]: Why was this file changed?

File: src/arithmetic/func_desc.c [hunk 35]
```diff
@@ -45,17 +33,17 @@ AR_FuncDesc *AR_FuncDescNew(const char *name, AR_Func func, uint min_argc, uint
...
-	_toLower(func->name, &lower_func_name[0], &lower_func_name_len);
+	size_t lower_func_name_len = 32;
+	str_tolower(func->name, lower_func_name, &lower_func_name_len);
 	assert(raxInsert(__aeRegisteredFuncs, (unsigned char *)lower_func_name, lower_func_name_len, func,
 					 NULL) == 1);
 }
 
 /* Get arithmetic function. */
 AR_FuncDesc *AR_GetFunc(const char *func_name) {
 	char lower_func_name[32] = {0};
```
Comment (swilly22) [inline src/arithmetic/func_desc.c:44]: This is a good opportunity to remove this initialisation
`char lower_func_name[32];`
Please apply throughout the modified files.

File: src/arithmetic/repository.c [hunk 36]
```diff
@@ -34,17 +22,17 @@ static void __agg_initRegistry() {
 void Agg_RegisterFunc(const char *name, AggFuncInit f) {
 	__agg_initRegistry();
 	char lower_func_name[32] = {0};
```
Comment (swilly22) [inline src/arithmetic/repository.c:24]: This is a good opportunity to remove this initialisation
char lower_func_name[32];
Please apply throughout the modified files.

File: src/procedures/procedure.c [hunk 37]
```diff
@@ -58,7 +63,11 @@ ProcedureCtx *ProcCtxNew(const char *name,
 
 ProcedureCtx *Proc_Get(const char *proc_name) {
 	if(!__procedures) return NULL;
-	ProcGenerator gen = raxFind(__procedures, (unsigned char *)proc_name, strlen(proc_name));
+	size_t proc_name_len = strlen(proc_name);
```
Comment (swilly22) [inline src/procedures/procedure.c:66]: `size_t proc_name_len = strlen(proc_name) + 1;`

**Review by swilly22 — APPROVED** [review]
Please see comments

**Review by swilly22 — APPROVED** [review]
